File: android/src/main/java/com/reactnativetimjs/manager/ConversationManager.java
package com.reactnativetimjs.manager;

import com.reactnativetimjs.util.CommonUtils;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReadableMap;

import com.tencent.imsdk.v2.V2TIMCallback;
import com.tencent.imsdk.v2.V2TIMConversation;
import com.tencent.imsdk.v2.V2TIMConversationListener;
import com.tencent.imsdk.v2.V2TIMConversationResult;
import com.tencent.imsdk.v2.V2TIMManager;
import com.tencent.imsdk.v2.V2TIMValueCallback;
import com.tencent.imsdk.v2.V2TIMConversationOperationResult;
import com.tencent.imsdk.v2.V2TIMConversationListFilter;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;

public class ConversationManager {
    private static V2TIMConversationListener conversationListener;

    public void removeConversationListener(Promise promise, ReadableMap arguments) {
        V2TIMManager.getConversationManager().removeConversationListener(conversationListener);
        CommonUtils.returnSuccess(promise, "removeConversationListener is done");
    }

    public void addConversationListener(Promise promise, ReadableMap arguments) {
        conversationListener = new V2TIMConversationListener() {
            @Override
            public void onSyncServerStart() {
                makeConversationListenerEventData("onSyncServerStart", null);
            }

            @Override
            public void onSyncServerFinish() {
                makeConversationListenerEventData("onSyncServerFinish", null);
            }

            @Override
            public void onSyncServerFailed() {
                makeConversationListenerEventData("onSyncServerFailed", null);
            }

            @Override
            public void onNewConversation(List<V2TIMConversation> conversationList) {
                LinkedList<HashMap<String, Object>> list = new LinkedList<HashMap<String, Object>>();
                for (int i = 0; i < conversationList.size(); i++) {
                    list.add(CommonUtils.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                makeConversationListenerEventData("onNewConversation", list);
            }

            @Override
            public void onConversationChanged(List<V2TIMConversation> conversationList) {
                LinkedList<HashMap<String, Object>> list = new LinkedList<HashMap<String, Object>>();
                for (int i = 0; i < conversationList.size(); i++) {
                    list.add(CommonUtils.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                makeConversationListenerEventData("onConversationChanged", list);
            }

            @Override
            public void onTotalUnreadMessageCountChanged(long totalUnreadCount) {
                makeConversationListenerEventData("onTotalUnreadMessageCountChanged", totalUnreadCount);
            }

            @Override
            public void onConversationGroupCreated(String groupName, List<V2TIMConversation> conversationList) {
                HashMap<String, Object> data = new HashMap<>();
                data.put("groupName", groupName);
                LinkedList<HashMap<String, Object>> list = new LinkedList<HashMap<String, Object>>();
                for (int i = 0; i < conversationList.size(); i++) {
                    list.add(CommonUtils.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                data.put("conversationList", list);
                makeConversationListenerEventData("onConversationGroupCreated", data);
            }

            @Override
            public void onConversationGroupDeleted(String groupName) {
                makeConversationListenerEventData("onConversationGroupDeleted", groupName);
            }

            @Override
            public void onConversationGroupNameChanged(String oldName, String newName) {
                HashMap<String, Object> data = new HashMap<>();
                data.put("oldName", oldName);
                data.put("newName", newName);
                makeConversationListenerEventData("onConversationGroupNameChanged", data);
            }

            @Override
            public void onConversationsAddedToGroup(String groupName, List<V2TIMConversation> conversationList) {
                HashMap<String, Object> data = new HashMap<>();
                data.put("groupName", groupName);
                LinkedList<HashMap<String, Object>> list = new LinkedList<HashMap<String, Object>>();
                for (int i = 0; i < conversationList.size(); i++) {
                    list.add(CommonUtils.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                data.put("conversationList", list);
                makeConversationListenerEventData("onConversationsAddedToGroup", data);
            }

            @Override
            public void onConversationsDeletedFromGroup(String groupName, List<V2TIMConversation> conversationList) {
                HashMap<String, Object> data = new HashMap<>();
                data.put("groupName", groupName);
                LinkedList<HashMap<String, Object>> list = new LinkedList<HashMap<String, Object>>();
                for (int i = 0; i < conversationList.size(); i++) {
                    list.add(CommonUtils.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                data.put("conversationList", list);
                makeConversationListenerEventData("onConversationsDeletedFromGroup", data);
            }
        };
        V2TIMManager.getConversationManager().addConversationListener(conversationListener);
        CommonUtils.returnSuccess(promise, "addConversationListener success");
    }

    public void getConversation(Promise promise, ReadableMap arguments) {
        String conversationID = arguments.getString("conversationID");
        V2TIMManager.getConversationManager().getConversation(conversationID,
                new V2TIMValueCallback<V2TIMConversation>() {
                    @Override
                    public void onError(int i, String s) {
                        CommonUtils.returnError(promise, i, s);
                    }

                    @Override
                    public void onSuccess(V2TIMConversation v2TIMConversation) {
                        CommonUtils.returnSuccess(promise,
                                CommonUtils.convertV2TIMConversationToMap(v2TIMConversation));
                    }
                });
    }

    private <T> void makeConversationListenerEventData(String eventType, T data) {
        CommonUtils.emmitEvent("conversationListener", eventType, data);
    }

    public void getConversationList(Promise promise, ReadableMap arguments) {
        String nextSeq = arguments.getString("nextSeq");
        int count = arguments.getInt("count");
        V2TIMManager.getConversationManager().getConversationList(Long.parseLong(nextSeq), count,
                new V2TIMValueCallback<V2TIMConversationResult>() {
                    @Override
                    public void onError(int i, String s) {
                        CommonUtils.returnError(promise, i, s);
                    }

                    @Override
                    public void onSuccess(V2TIMConversationResult v2TIMConversationResult) {
                        CommonUtils.returnSuccess(promise,
                                CommonUtils.convertV2TIMConversationResultToMap(v2TIMConversationResult));
                    }
                });
    }

    public void getConversationListByConversaionIds(Promise promise, ReadableMap arguments) {
        List<String> conversationIDList = CommonUtils
                .convertReadableArrayToListString(arguments.getArray("conversationIDList"));
        V2TIMManager.getConversationManager().getConversationList(conversationIDList,
                new V2TIMValueCallback<List<V2TIMConversation>>() {
                    @Override
                    public void onSuccess(List<V2TIMConversation> v2TIMConversations) {
                        LinkedList<HashMap<String, Object>> list = new LinkedList<HashMap<String, Object>>();
                        for (int i = 0; i < v2TIMConversations.size(); i++) {
                            list.add(CommonUtils.convertV2TIMConversationToMap(v2TIMConversations.get(i)));
                        }
                        CommonUtils.returnSuccess(promise, list);
                    }

                    @Override
                    public void onError(int i, String s) {
                        CommonUtils.returnError(promise, i, s);
                    }
                });
    }

    public void pinConversation(Promise promise, ReadableMap arguments) {
        String conversationID = arguments.getString("conversationID");
        boolean isPinned = arguments.getBoolean("isPinned");
        V2TIMManager.getConversationManager().pinConversation(conversationID, isPinned, new V2TIMCallback() {
            @Override
            public void onSuccess() {
                CommonUtils.returnSuccess(promise, null);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtils.returnError(promise, i, s);
            }
        });
    }

    public void getTotalUnreadMessageCount(Promise promise, ReadableMap arguments) {
        V2TIMManager.getConversationManager().getTotalUnreadMessageCount(new V2TIMValueCallback<Long>() {
            @Override
            public void onSuccess(Long aLong) {
                CommonUtils.returnSuccess(promise, aLong);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtils.returnError(promise, i, s);
            }
        });
    }

    public void deleteConversation(Promise promise, ReadableMap arguments) {
        // session id
        String conversationID = arguments.getString("conversationID");
        V2TIMManager.getConversationManager().deleteConversation(conversationID, new V2TIMCallback() {
            @Override
            public void onError(int i, String s) {
                CommonUtils.returnError(promise, i, s);
            }

            @Override
            public void onSuccess() {
                CommonUtils.returnSuccess(promise, null);
            }
        });

    }

    public void setConversationDraft(Promise promise, ReadableMap arguments) {
        // session id
        String conversationID = arguments.getString("conversationID");
        String draftText = arguments.getString("draftText");
        if (draftText == "") {
            draftText = null;
        }
        V2TIMManager.getConversationManager().setConversationDraft(conversationID, draftText, new V2TIMCallback() {
            @Override
            public void onError(int i, String s) {
                CommonUtils.returnError(promise, i, s);
            }

            @Override
            public void onSuccess() {
                CommonUtils.returnSuccess(promise, null);
            }
        });

    }

    public void addConversationsToGroup(Promise promise, ReadableMap arguments) {
        List<String> conversationIDList = CommonUtils
                .convertReadableArrayToListString(arguments.getArray("conversationIDList"));
        String groupName = arguments.getString("groupName");
        V2TIMManager.getConversationManager().addConversationsToGroup(groupName, conversationIDList,
                new V2TIMValueCallback<List<V2TIMConversationOperationResult>>() {
                    @Override
                    public void onSuccess(List<V2TIMConversationOperationResult> v2TIMConversationOperationResults) {
                        LinkedList<HashMap<String, Object>> list = new LinkedList<HashMap<String, Object>>();
                        for (int i = 0; i < v2TIMConversationOperationResults.size(); i++) {
                            list.add(CommonUtils.convertV2TIMConversationOperationResultToMap(
                                    v2TIMConversationOperationResults.get(i)));
                        }
                        CommonUtils.returnSuccess(promise, list);
                    }

                    @Override
                    public void onError(int i, String s) {
                        CommonUtils.returnError(promise, i, s);
                    }
                });
    }

    public void deleteConversationsFromGroup(Promise promise, ReadableMap arguments) {
        List<String> conversationIDList = CommonUtils
                .convertReadableArrayToListString(arguments.getArray("conversationIDList"));
        String groupName = arguments.getString("groupName");
        V2TIMManager.getConversationManager().deleteConversationsFromGroup(groupName, conversationIDList,
                new V2TIMValueCallback<List<V2TIMConversationOperationResult>>() {
                    @Override
                    public void onSuccess(List<V2TIMConversationOperationResult> v2TIMConversationOperationResults) {
                        LinkedList<HashMap<String, Object>> list = new LinkedList<HashMap<String, Object>>();
                        for (int i = 0; i < v2TIMConversationOperationResults.size(); i++) {
                            list.add(CommonUtils.convertV2TIMConversationOperationResultToMap(
                                    v2TIMConversationOperationResults.get(i)));
                        }
                        CommonUtils.returnSuccess(promise, list);
                    }

                    @Override
                    public void onError(int i, String s) {
                        CommonUtils.returnError(promise, i, s);
                    }
                });
    }

    public void setConversationCustomData(Promise promise, ReadableMap arguments) {
        List<String> conversationIDList = CommonUtils
                .convertReadableArrayToListString(arguments.getArray("conversationIDList"));
        String customData = arguments.getString("customData");
        V2TIMManager.getConversationManager().setConversationCustomData(conversationIDList, customData,
                new V2TIMValueCallback<List<V2TIMConversationOperationResult>>() {
                    @Override
                    public void onSuccess(List<V2TIMConversationOperationResult> v2TIMConversationOperationResults) {
                        LinkedList<HashMap<String, Object>> list = new LinkedList<>();
                        for (int i = 0; i < v2TIMConversationOperationResults.size(); i++) {
                            list.add(CommonUtils.convertV2TIMConversationOperationResultToMap(
                                    v2TIMConversationOperationResults.get(i)));
                        }
                        CommonUtils.returnSuccess(promise, list);
                    }

                    @Override
                    public void onError(int i, String s) {
                        CommonUtils.returnError(promise, i, s);
                    }
                });
    }

    public void getConversationListByFilter(Promise promise, ReadableMap arguments) {
        final HashMap<String, Object> filter = CommonUtils
                .convertReadableMapToHashMap(CommonUtils.safeGetMap(arguments, "filter"));
        final V2TIMConversationListFilter filterNative = new V2TIMConversationListFilter();
        if (filter.get("conversationType") != null) {
            filterNative.setConversationType((Integer) filter.get("conversationType"));
        }
        if (filter.get("nextSeq") != null) {
            filterNative.setNextSeq(Long.parseLong(filter.get("nextSeq").toString()));
        }
        if (filter.get("count") != null) {
            filterNative.setCount((Integer) filter.get("count"));
        }
        if (filter.get("markType") != null) {
            filterNative.setMarkType(0x1l << (int) filter.get("markType"));
        }
        if (filter.get("groupName") != null) {
            filterNative.setGroupName((String) filter.get("groupName"));
        }
        V2TIMManager.getConversationManager().getConversationListByFilter(filterNative,
                new V2TIMValueCallback<V2TIMConversationResult>() {
                    @Override
                    public void onSuccess(V2TIMConversationResult v2TIMConversationResult) {
                        CommonUtils.returnSuccess(promise,
                                CommonUtils.convertV2TIMConversationResultToMap(v2TIMConversationResult));
                    }

                    @Override
                    public void onError(int i, String s) {
                        CommonUtils.returnError(promise, i, s);
                    }
                });

    }

    public void markConversation(Promise promise, ReadableMap arguments) {
        List<String> conversationIDList = CommonUtils
                .convertReadableArrayToListString(arguments.getArray("conversationIDList"));
        boolean enableMark = arguments.getBoolean("enableMark");
        int markType = arguments.getInt("markType");
        long longMarkType = 0x1l << markType;
        V2TIMManager.getConversationManager().markConversation(conversationIDList, longMarkType, enableMark,
                new V2TIMValueCallback<List<V2TIMConversationOperationResult>>() {
                    @Override
                    public void onSuccess(List<V2TIMConversationOperationResult> v2TIMConversationOperationResults) {
                        LinkedList<HashMap<String, Object>> list = new LinkedList<HashMap<String, Object>>();
                        for (int i = 0; i < v2TIMConversationOperationResults.size(); i++) {
                            list.add(CommonUtils.convertV2TIMConversationOperationResultToMap(
                                    v2TIMConversationOperationResults.get(i)));
                        }
                        CommonUtils.returnSuccess(promise, list);
                    }

                    @Override
                    public void onError(int i, String s) {
                        CommonUtils.returnError(promise, i, s);
                    }
                });

    }

    public void createConversationGroup(Promise promise, ReadableMap arguments) {
        List<String> conversationIDList = CommonUtils
                .convertReadableArrayToListString(arguments.getArray("conversationIDList"));
        String groupName = arguments.getString("groupName");
        V2TIMManager.getConversationManager().createConversationGroup(groupName, conversationIDList,
                new V2TIMValueCallback<List<V2TIMConversationOperationResult>>() {
                    @Override
                    public void onSuccess(List<V2TIMConversationOperationResult> v2TIMConversationOperationResults) {
                        LinkedList<HashMap<String, Object>> list = new LinkedList<HashMap<String, Object>>();
                        for (int i = 0; i < v2TIMConversationOperationResults.size(); i++) {
                            list.add(CommonUtils.convertV2TIMConversationOperationResultToMap(
                                    v2TIMConversationOperationResults.get(i)));
                        }
                        CommonUtils.returnSuccess(promise, list);
                    }

                    @Override
                    public void onError(int i, String s) {
                        CommonUtils.returnError(promise, i, s);
                    }
                });
    }

    public void getConversationGroupList(Promise promise, ReadableMap arguments) {
        V2TIMManager.getConversationManager().getConversationGroupList(new V2TIMValueCallback<List<String>>() {
            @Override
            public void onSuccess(List<String> strings) {
                List<String> list = new LinkedList();
                for (int i = 0; i < strings.size(); i++) {
                    list.add(strings.get(i));
                }
                System.out.println(list);
                CommonUtils.returnSuccess(promise, list);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtils.returnError(promise, i, s);
            }
        });
    }

    public void deleteConversationGroup(Promise promise, ReadableMap arguments) {
        String groupName = arguments.getString("groupName");
        V2TIMManager.getConversationManager().deleteConversationGroup(groupName, new V2TIMCallback() {
            @Override
            public void onSuccess() {
                CommonUtils.returnSuccess(promise, null);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtils.returnError(promise, i, s);
            }
        });
    }

    public void renameConversationGroup(Promise promise, ReadableMap arguments) {
        String oldName = arguments.getString("oldName");
        String newName = arguments.getString("newName");
        V2TIMManager.getConversationManager().renameConversationGroup(oldName, newName, new V2TIMCallback() {
            @Override
            public void onSuccess() {
                CommonUtils.returnSuccess(promise, null);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtils.returnError(promise, i, s);
            }
        });

    }
}

File: ios/manager/conversationManager.swift
import Foundation
import ImSDK_Plus

class ConversationManager {
	private var conversationListener: ConversationListener = ConversationListener();

	public func getConversationList(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		if let nextSeq = param["nextSeq"] as? String,
		   let count = param["count"] as? Int32 {
            V2TIMManager.sharedInstance().getConversationList(UInt64(nextSeq) ?? 0 , count: count, succ: {
				conversations, nextSeq, finished in
				
                let dict = V2ConversationResultEntity.init(conversations: conversations ?? [], nextSeq: String(nextSeq), finished: finished).getDict();
				CommonUtils.resultSuccess(method: "getConversationList", resolve: resolve, data: dict);
			}, fail: TencentImUtils.returnErrorClosures(method: "getConversationList", resolve: resolve))
		}
	}
	
	public func getConversationListByConversaionIds(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		if let conversationIDList =  param["conversationIDList"] as? [String] {
		   V2TIMManager.sharedInstance().getConversationList(conversationIDList, succ: {
               if($0 != nil){
                   let dict = $0!.map { V2ConversationEntity.getDict(info: $0) };
                   CommonUtils.resultSuccess(method: "getConversationListByConversaionIds", resolve: resolve, data: dict);
               }
				
			}, fail: TencentImUtils.returnErrorClosures(method: "getConversationListByConversaionIds", resolve: resolve))
		}
	}
	
	public func getConversation(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		if let conversationID = param["conversationID"] as? String {
			V2TIMManager.sharedInstance().getConversation(conversationID, succ: {
				conversation in
				
				CommonUtils.resultSuccess(method: "getConversation", resolve: resolve, data: V2ConversationEntity.getDict(info: conversation!));
			}, fail: TencentImUtils.returnErrorClosures(method: "getConversation", resolve: resolve))
		}
	}
	
	public func deleteConversation(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		let conversationID = param["conversationID"] as? String;
		
		V2TIMManager.sharedInstance()?.deleteConversation(conversationID, succ: {
			() -> Void in
			
			CommonUtils.resultSuccess(method: "deleteConversation", resolve: resolve)
		}, fail: TencentImUtils.returnErrorClosures(method: "deleteConversation", resolve: resolve));
	}
	
	public func setConversationDraft(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		let conversationID = param["conversationID"] as? String;
		let draftText = param["draftText"] as? String;
		
		V2TIMManager.sharedInstance()?.setConversationDraft(conversationID, draftText: draftText, succ: {
			() -> Void in
			
			CommonUtils.resultSuccess(method: "setConversationDraft", resolve: resolve)
		}, fail: TencentImUtils.returnErrorClosures(method: "setConversationDraft", resolve: resolve))
	}
	
	public func pinConversation(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		let conversationID = param["conversationID"] as? String;
		let isPinned = param["isPinned"] as? Bool;
		
		V2TIMManager.sharedInstance()?.pinConversation(conversationID, isPinned: isPinned ?? false, succ: {
			() -> Void in
			CommonUtils.resultSuccess(method: "pinConversation", resolve: resolve, data: "ok")
		}, fail: TencentImUtils.returnErrorClosures(method: "pinConversation", resolve: resolve))
	}
	
	public func getTotalUnreadMessageCount(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		V2TIMManager.sharedInstance()?.getTotalUnreadMessageCount({
			totalCount -> Void in
			CommonUtils.resultSuccess(method: "getTotalUnreadMessageCount", resolve: resolve, data: totalCount)
		}, fail: TencentImUtils.returnErrorClosures(method: "getTotalUnreadMessageCount", resolve: resolve))
	}

	public func addConversationListener(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
        V2TIMManager.sharedInstance().addConversationListener(listener: conversationListener);
		CommonUtils.resultSuccess(method: "addConversationListener", resolve: resolve, data: "addConversationListener is done");
	}

	public func removeConversationListener(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		V2TIMManager.sharedInstance().removeConversationListener(listener: conversationListener);
		CommonUtils.resultSuccess(method: "removeConversationListener", resolve: resolve, data: "removeConversationListener is done");
	}

	public func addConversationsToGroup(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
        let conversationIDList:[String] = param["conversationIDList"] as! [String];
        let groupName = (param["groupName"] as? String)!;
        V2TIMManager.sharedInstance().addConversations(toGroup: groupName, conversationIDList: conversationIDList) { result_list in
            var dd : [[String:Any]] = [];
            result_list?.forEach({ item in
                dd.append([
                    "conversationID":item.conversationID ?? "",
                    "resultCode":item.resultCode ,
                    "resultInfo":item.resultInfo ?? "",
                ])
            })
            CommonUtils.resultSuccess(method: "addConversationsToGroup", resolve: resolve, data: dd)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, method: "addConversationsToGroup", resolve: resolve)
        }
    }

	public func deleteConversationsFromGroup(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
        let conversationIDList:[String] = param["conversationIDList"] as! [String];
        let groupName = (param["groupName"] as? String)!;
        V2TIMManager.sharedInstance().deleteConversations(fromGroup: groupName, conversationIDList: conversationIDList) { result_list in
            var dd : [[String:Any]] = [];
            result_list?.forEach({ item in
                dd.append([
                    "conversationID":item.conversationID ?? "",
                    "resultCode":item.resultCode ,
                    "resultInfo":item.resultInfo ?? "",
                ])
            })
			CommonUtils.resultSuccess(method: "deleteConversationsFromGroup", resolve: resolve, data: dd)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, method: "deleteConversationsFromGroup", resolve: resolve)
        }
        
    }

	public func setConversationCustomData(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		let conversationIDList:[String] = param["conversationIDList"] as! [String];
        let customData = (param["customData"] as? String)!;
        V2TIMManager.sharedInstance().setConversationCustomData(conversationIDList, customData: customData.data(using: String.Encoding.utf8)) { result_list in
            var dd : [[String:Any]] = [];
            result_list?.forEach({ item in
                dd.append([
                    "conversationID":item.conversationID ?? "",
                    "resultCode":item.resultCode ,
                    "resultInfo":item.resultInfo ?? "",
                ])
            })
            CommonUtils.resultSuccess(method: "setConversationCustomData", resolve: resolve, data: dd)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, method: "setConversationCustomData", resolve: resolve)
        }

    }

    public func getConversationListByFilter(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
        let filter = param["filter"] as? [String:Any] ?? [:];
        let filter_native = V2TIMConversationListFilter.init();
      
        if(filter["groupName"] as? String != nil){
            filter_native.groupName = (filter["groupName"] as? String)
        }
        
        if(filter["conversationType"]  as? Int != nil){
            filter_native.type = V2TIMConversationType.init(rawValue: Int(filter["conversationType"] as! Int)) ?? V2TIMConversationType.C2C
        }

        if(filter["nextSeq"]  as? UInt64 != nil){
            filter_native.nextSeq = (filter["nextSeq"] as! UInt64)
        }
        if(filter["count"]  as? UInt32 != nil){
            filter_native.count = (filter["count"] as! UInt)
        }
        if(filter["markType"]  as? UInt32 != nil){
            let u64: UInt64 = 1;
            let ab = u64 << (filter["markType"] as! UInt32)
            filter_native.markType = UInt(ab);
        }
        V2TIMManager.sharedInstance().getConversationList(by: filter_native) { conversations, nextSeq, finished in
            let dict = V2ConversationResultEntity.init(conversations: conversations ?? [], nextSeq: String(nextSeq), finished: finished).getDict();
            CommonUtils.resultSuccess(method: "getConversationListByFilter", resolve: resolve, data: dict);
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, method: "getConversationListByFilter", resolve: resolve)
        }

    }

    public func markConversation(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		let conversationIDList:[String] = param["conversationIDList"] as! [String];
        let enableMark = (param["enableMark"] as? Bool)!;
        let markType = (param["markType"] as? Int32)!;
        let u64: UInt64 = 1;
        V2TIMManager.sharedInstance().markConversation(conversationIDList, markType: (u64 << markType) as NSNumber?, enableMark: enableMark) { result_list in
            var dd : [[String:Any]] = [];
            result_list?.forEach({ item in
                dd.append([
                    "conversationID":item.conversationID ?? "",
                    "resultCode":item.resultCode ,
                    "resultInfo":item.resultInfo ?? "",
                ])
            })
            CommonUtils.resultSuccess(method: "markConversation", resolve: resolve, data: dd)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, method: "markConversation", resolve: resolve)
        }
    }

    public func createConversationGroup(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		let conversationIDList:[String] = param["conversationIDList"] as! [String];
        let groupName = (param["groupName"] as? String)!;
        V2TIMManager.sharedInstance().createConversationGroup(groupName, conversationIDList: conversationIDList) { result_list in
            var dd : [[String:Any]] = [];
            result_list?.forEach({ item in
                dd.append([
                    "conversationID":item.conversationID ?? "",
                    "resultCode":item.resultCode ,
                    "resultInfo":item.resultInfo ?? "",
                ])
            })
            CommonUtils.resultSuccess(method: "createConversationGroup", resolve: resolve, data: dd)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, method: "createConversationGroup", resolve: resolve)
        }
    }
    
    public func getConversationGroupList(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
        V2TIMManager.sharedInstance().getConversationGroupList { ids in
            CommonUtils.resultSuccess(method: "getConversationGroupList", resolve: resolve, data: ids ?? [])
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, method: "getConversationGroupList", resolve: resolve)
        }

    }
    public func deleteConversationGroup(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
        let groupName = (param["groupName"] as? String)!;
        V2TIMManager.sharedInstance().deleteConversationGroup(groupName) {
            CommonUtils.resultSuccess(method: "deleteConversationGroup", resolve: resolve)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, method: "deleteConversationGroup", resolve: resolve)
        }

    }
    public func renameConversationGroup(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
        let oldName = (param["oldName"] as? String)!;
        let newName = (param["newName"] as? String)!;
        V2TIMManager.sharedInstance().renameConversationGroup(oldName, newName: newName) {
            CommonUtils.resultSuccess(method: "renameConversationGroup", resolve: resolve)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, method: "renameConversationGroup", resolve: resolve)
        }

    }
}
